packages/react-components/react-timepicker-compat-preview/src/components/TimePicker/useTimePicker.tsx

```diff
@@ -202,5 +201,30 @@ const useSelectTimeFromValue = (state: TimePickerState, callback: TimePickerProp
   );
   state.root.onKeyDown = mergeCallbacks(handleKeyDown, state.root.onKeyDown);
 
-  // TODO call selectTimeFromValue on blur
+  const rootRef = React.useRef<HTMLDivElement>(null);
```

Comment:
Since there the listbox is normally rendered in a portal, you can simply import `elementContains` from `@fluentui/react-utilities` and use that with only the root ref.

```ts
const useSelectTimeFromValue = (state: TimePickerState, callback: TimePickerProps['onTimeSelect']) => {
  const { activeOption, freeform, validateFreeFormTime, options, submittedText, setActiveOption, value } = state;

  // Base Combobox has activeOption default to first option in dropdown even if it doesn't match input value, and Enter key will select it.
  // This effect ensures that the activeOption is cleared when the input doesn't match any option.
  // This behavior is specific to a freeform TimePicker where the input value is treated as a valid time even if it's not in the dropdown.
  const isValueOptionPrefix = value ? options.some(({ text }) => text.indexOf(value) === 0) : false;
  React.useEffect(() => {
    if (freeform && value && !isValueOptionPrefix) {
      setActiveOption(undefined);
    }
  }, [freeform, isValueOptionPrefix, setActiveOption, value]);

  const selectTimeFromValue = React.useCallback(
    (e: React.KeyboardEvent<HTMLInputElement> | React.FocusEvent<HTMLInputElement>) => {
      if (!freeform) {
        return;
      }

      const { date: selectedTime, error } = validateFreeFormTime(value);

      // Only triggers callback when the text in input has changed.
      if (submittedText !== value) {
        callback?.(e, { selectedTime, selectedTimeText: value, error });
      }
    },
    [callback, freeform, submittedText, validateFreeFormTime, value],
  );

  const handleKeyDown: ComboboxProps['onKeyDown'] = React.useCallback(
    e => {
      if (!activeOption && e.key === Enter) {
        selectTimeFromValue(e);
      }
    },
    [activeOption, selectTimeFromValue],
  );
  state.root.onKeyDown = mergeCallbacks(handleKeyDown, state.root.onKeyDown);

  const rootRef = React.useRef<HTMLDivElement>(null);
  state.root.ref = useMergedRefs(state.root.ref, rootRef);

  if (state.listbox) {
    state.listbox.tabIndex = -1; // allows it to be the relatedTarget of a blur event.
  }

  if (state.expandIcon) {
    state.expandIcon.tabIndex = -1; // allows it to be the relatedTarget of a blur event.
  }

  const handleInputBlur = React.useCallback(
    (e: React.FocusEvent<HTMLInputElement>) => {
      const isOutside = e.relatedTarget ? !elementContains(rootRef.current, e.relatedTarget) : true;
      if (isOutside) {
        selectTimeFromValue(e);
      }
    },
    [selectTimeFromValue],
  );
  state.input.onBlur = mergeCallbacks(handleInputBlur, state.input.onBlur);
};

```

Reply:
love it, updated :)